### PR TITLE
[native bridge] - move compiler, execution changes for bridge

### PR DESCRIPTION
**Pull request by patrickkuo**
## Description 

Adding move compiler object id and OTW exception for bridge prior adding to the framework package 
## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-types/src/bridge.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+use crate::object::Owner;
+use crate::storage::ObjectStore;
+use crate::SUI_BRIDGE_OBJECT_ID;
+use move_core_types::ident_str;
+use move_core_types::identifier::IdentStr;
+
+pub const BRIDGE_MODULE_NAME: &IdentStr = ident_str!("bridge");
+pub const BRIDGE_CREATE_FUNCTION_NAME: &IdentStr = ident_str!("create");
+
+pub fn get_bridge_obj_initial_shared_version(
```
Comment (tnowacki) [inline crates/sui-types/src/bridge.rs:15]: For clock, this is statically knowable. Is this not the case for this bridge object?

File: external-crates/move/crates/move-compiler/src/sui_mode/typing.rs [hunk 2]
```diff
@@ -916,6 +913,18 @@ fn exp(context: &mut Context, e: &T::Exp) {
     }
 }
 
+fn otw_special_cases(context: &Context) -> bool {
+    for token in ["btc", "eth", "usdc", "usdt"] {
+        if context.current_module().value.is(BRIDGE_ADDR_NAME, token) {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/sui_mode/typing.rs:918]: nit, just a bit more idiomatic 
`["btc", "eth", "usdc", "usdt"].iter().any(|token| context.current_module().value.is(BRIDGE_ADDR_NAME, token))`

File: sui-execution/latest/sui-verifier/src/one_time_witness_verifier.rs [hunk 3]
```diff
@@ -49,6 +49,12 @@ pub fn verify_module(
         return Ok(());
     }
 
+    for token in ["btc", "eth", "usdc", "usdt"] {
+        if ModuleId::new(BRIDGE_ADDRESS, ident_str!(token).to_owned()) == module.self_id() {
```
Comment (tnowacki) [inline sui-execution/latest/sui-verifier/src/one_time_witness_verifier.rs:53]: Just a nit same as above, `iter().any(...)`  is a bit more idiomatic

File: sui-execution/latest/sui-verifier/src/one_time_witness_verifier.rs [hunk 4]
```diff
@@ -49,6 +49,12 @@ pub fn verify_module(
         return Ok(());
     }
 
+    for token in ["btc", "eth", "usdc", "usdt"] {
```
Comment (tnowacki) [inline sui-execution/latest/sui-verifier/src/one_time_witness_verifier.rs:52]: Nit, maybe this to a constant

File: external-crates/move/crates/move-compiler/src/sui_mode/typing.rs [hunk 5]
```diff
@@ -916,6 +913,18 @@ fn exp(context: &mut Context, e: &T::Exp) {
     }
 }
 
+fn otw_special_cases(context: &Context) -> bool {
+    for token in ["btc", "eth", "usdc", "usdt"] {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/sui_mode/typing.rs:917]: nit, bind this to a constant

File: sui-execution/latest/sui-verifier/src/one_time_witness_verifier.rs (hunk 3 above)
Comment (tnowacki) [reply sui-execution/latest/sui-verifier/src/one_time_witness_verifier.rs:53]: nit, calling `self_id()` allocates a new Identifier, we should bind it and share it for all these checks above

File: sui-execution/v1/sui-verifier/src/one_time_witness_verifier.rs [hunk 6]
```diff
@@ -28,7 +28,7 @@ use sui_types::{
     base_types::{TX_CONTEXT_MODULE_NAME, TX_CONTEXT_STRUCT_NAME},
     error::ExecutionError,
     move_package::{is_test_fun, FnInfoMap},
-    SUI_FRAMEWORK_ADDRESS,
+    BRIDGE_ADDRESS, SUI_FRAMEWORK_ADDRESS,
```
Comment (tnowacki) [inline sui-execution/v1/sui-verifier/src/one_time_witness_verifier.rs:31]: Should we be changing this in v1 and v0? I feel like we shouldn't need to but maybe I'm missing something

File: crates/sui-types/src/lib.rs [hunk 7]
```diff
@@ -128,12 +134,20 @@ pub const SUI_RANDOMNESS_STATE_OBJECT_ID: ObjectID =
 pub const SUI_DENY_LIST_ADDRESS: AccountAddress = deny_list_addr();
 pub const SUI_DENY_LIST_OBJECT_ID: ObjectID = ObjectID::from_address(SUI_DENY_LIST_ADDRESS);
 
+/// 0x9: hardcode object ID for the singleton bridge object.
+pub const SUI_BRIDGE_ADDRESS: AccountAddress = address_from_single_byte(9);
+pub const SUI_BRIDGE_OBJECT_ID: ObjectID = ObjectID::from_address(SUI_BRIDGE_ADDRESS);
```
Comment (tnowacki) [inline crates/sui-types/src/lib.rs:139]: I think we will need some more checks on the bridge object:
- Can we take it mutably? 
  - If no, we will need to duplicate the check like we have for other system objects 
  - If yes, make sure it doesn't have `store` so no one can wrap it or steal it

File: sui-execution/latest/sui-verifier/src/one_time_witness_verifier.rs (hunk 3 above)
Comment (tnowacki) [reply sui-execution/latest/sui-verifier/src/one_time_witness_verifier.rs:53]: And if we want, we can avoid checking the address n-times by checking it first outside the loop

File: crates/sui-types/src/bridge.rs (hunk 1 above)
Comment (patrickkuo) [reply crates/sui-types/src/bridge.rs:15]: The bridge object is created in a End of epoch transaction, so the version is unknown.

File: sui-execution/v1/sui-verifier/src/one_time_witness_verifier.rs (hunk 6 above)
Comment (patrickkuo) [reply sui-execution/v1/sui-verifier/src/one_time_witness_verifier.rs:31]: you are right, we don;t need to change v0 and v1

File: crates/sui-types/src/lib.rs (hunk 7 above)
Comment (patrickkuo) [reply crates/sui-types/src/lib.rs:139]: the Bridge object is mutable, it is versioned and the wrapper object only have `key` ability

**Review by tnowacki — APPROVED** [review]
LGTM

**Pull request merged**